services/ingestion/etl/fts/xls/src/lib/transform.js

```diff
@@ -0,0 +1,234 @@
+// @flow
+
+import countries from 'i18n-iso-countries';
+import type { Project } from '@eubfr/types';
+import getCountryCode from '@eubfr/lib/getCountryCode';
+import sanitizeBudgetItem from '@eubfr/lib/budgetFormatter';
+
+/**
+ * Gets country code from a country name.
+ * Contains a map of exceptions.
+ *
+ * @memberof FtsXlsTransform
+ * @param {String} countryName The name of the country
+ * @returns {String} The ISO 3166-1 country code
+ */
+const getCodeByCountry = countryName => {
+  // There are a few exceptions which should be normalized, regardless of the library
+  const corrections = new Map([
+    ['Tanzania', 'Tanzania, United Republic of'],
+    ['Congo (Democratic Republic of)', 'Congo, the Democratic Republic of the'],
+    ['Serbia (Republic of)', 'Serbia'],
+    ['Occupied Territories - Palestine', 'Palestinian Territory, Occupied'],
+    [
+      'Macedonia (the former Yugoslav Republic of)',
+      'Macedonia, the Former Yugoslav Republic of',
+    ],
+  ]);
+
+  const countryNameCorrected = corrections.get(countryName) || countryName;
+
+  return countries.getAlpha2Code(countryNameCorrected, 'en');
+};
+
+/**
+ * Gets NUTS code level from a string
+ *
+ * @memberof FtsXlsTransform
+ * @param {String} code The NUTS code
+ * @returns {Number} The level of NUTS or null if one can't be extracted
+ */
+const getNutsCodeLevel = code => {
+  if (code && code.length >= 2) {
+    return code.length - 2;
+  }
+  return null;
+};
+
+/**
+ * Preprocess budget
+ *
+ * Input fields taken from the `record` are:
+ * - `Total amount`
+ * - `Funding Type`
+ * - `Type`
+ *
+ * @memberof FtsXlsTransform
+ * @param {Object} record The row received from parsed file
+ * @returns {Budget}
+ */
+const getBudget = record => {
+  const fundingArea = [];
+  // Use `Type` field if provided.
+  if (record.Type) {
+    fundingArea.push(record.Type);
+  }
+
+  // Use `Funding Type` field if provided.
+  if (record['Funding Type']) {
+    fundingArea.push(record['Funding Type']);
+  }
+
+  const budget = {
+    eu_contrib: sanitizeBudgetItem({
+      value: record['Total amount'],
+      currency: 'EUR',
+      raw: record['Total amount'],
+    }),
+    funding_area: fundingArea,
+    mmf_heading: '',
+    other_contrib: sanitizeBudgetItem(),
+    private_fund: sanitizeBudgetItem(),
+    public_fund: sanitizeBudgetItem(),
+    total_cost: sanitizeBudgetItem(),
+  };
+
+  return budget;
+};
+
+/**
+ * Preprocess project_locations
+ * Input fields taken from the `record` are:
+ * - `Address`
+ * - `City`
+ * - `Postal code`
+ * - `Country / Territory`
+ * - `NUTS2`
+ * - `Geographical Zone`
+ *
+ * @memberof FtsXlsTransform
+ * @param {Object} record The row received from parsed file
+ * @returns {Array} List of {Location} objects for `project_locations` field
+ */
+const getLocations = record => {
+  const locations = [];
+  const addressItems = record.Address.split(';');
```

Comment:
The code here is not really safe. If `record` doesn't have an `Address` field, then it will throw an error. It can be a feature ("if the document is not well-formattedn then throw an error") but it raises some concerns too.

A safe `get` (_à la_ https://lodash.com/docs/4.17.10#get) would prevent this issue.

What was the consensus about the schema validation?

Reply:
Reworked this one to include values in the specific fields only when the field is available in the source.